### PR TITLE
HIVE-25091: Implemented connector provider for MSSQL and Oracle

### PR DESCRIPTION
**Pull request by saihemanth-cloudera**
### What changes were proposed in this pull request?
Added implementations of connector provider for MSSQL and Oracle.

### Why are the changes needed?
WIth these changes users will be able to create connectors for mssql and oracle.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Local machine

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java [hunk 1]
```diff
@@ -172,10 +172,12 @@ protected Connection getConnection() {
     ResultSet rs = null;
     Table table = null;
     try {
-      // rs = fetchTableMetadata(tableName);
-      rs = fetchTableViaDBMetaData(tableName);
+      rs = fetchTableMetadata(tableName);
```
Comment (nrg4878) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java:175]: I am not sure I understand this change. Dont we want to use fetchTableViaDBmetadata? we comment out the implementation in the abstract class and moved it into subclasses. can you please expand on your changes.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java [hunk 2]
```diff
@@ -216,16 +218,16 @@ protected Connection getConnection() {
...
+//  private ResultSet fetchTableViaDBMetaData(String tableName) throws SQLException {
+//    ResultSet rs = null;
+//    try {
+//      rs = getConnection().getMetaData().getColumns(scoped_db, null, tableName, null);
+//    } catch (SQLException sqle) {
+//      LOG.warn("Could not retrieve column names from JDBC table, cause:" + sqle.getMessage());
+//      throw sqle;
+//    }
+//    return rs;
+//  }
```
Comment (belugabehr) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java:230]: Just drop the code if need be, do not check-in commented out code

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/DerbySQLConnectorProvider.java [hunk 3]
```diff
@@ -48,7 +48,14 @@ protected ResultSet fetchTableNames() throws MetaException {
    */
   @Override
   public ResultSet fetchTableMetadata(String tableName) throws MetaException {
-     return null;
+    ResultSet rs = null;
+    try {
+      rs = getConnection().getMetaData().getColumns(scoped_db, null, tableName, null);
+    } catch (SQLException sqle) {
+      LOG.warn("Could not retrieve column names from JDBC table, cause:" + sqle.getMessage());
```
Comment (belugabehr) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/DerbySQLConnectorProvider.java:55]: `LOG.warn("Could not retrieve column names from JDBC table", sqle);`

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/MSSQLConnectorProvider.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    @Override public List<Table> getTables(String regex) throws MetaException {
+        return null;
+    }
+
+    @Override protected ResultSet fetchTableMetadata(String tableName) throws MetaException {
+        ResultSet rs = null;
+        try {
+            rs = getConnection().getMetaData().getColumns(null, scoped_db, tableName, null);
+        } catch (SQLException sqle) {
+            LOG.warn("Could not retrieve table names from remote datasource, cause:" + sqle.getMessage());
```
Comment (belugabehr) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/MSSQLConnectorProvider.java:39]: `LOG.warn("Could not retrieve table names from JDBC table", sqle);`

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/MSSQLConnectorProvider.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            rs = getConnection().getMetaData().getTables(null, scoped_db, null, null);
+            if (rs != null) {
+                List<String> tables = new ArrayList<String>();
+                while(rs.next()) {
+                    tables.add(rs.getString(3));
+                }
+                return tables;
+            }
+        } catch (SQLException sqle) {
+            LOG.warn("Could not retrieve table names from remote datasource, cause:" + sqle.getMessage());
```
Comment (belugabehr) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/MSSQLConnectorProvider.java:61]: `LOG.warn("Failed to retrieve ....", sqle);`

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/MSSQLConnectorProvider.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    @Override protected ResultSet fetchTableNames() throws MetaException {
+        return null;
+    }
+
+    @Override public List<String> getTableNames() throws MetaException {
+        ResultSet rs = null;
+        try {
+            rs = getConnection().getMetaData().getTables(null, scoped_db, null, null);
+            if (rs != null) {
+                List<String> tables = new ArrayList<String>();
```
Comment (belugabehr) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/MSSQLConnectorProvider.java:54]: `= newArrayList<>();`

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/MSSQLConnectorProvider.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        } catch (SQLException sqle) {
+            LOG.warn("Could not retrieve table names from remote datasource, cause:" + sqle.getMessage());
+        } finally {
+            try {
+                if (rs != null) {
+                    rs.close();
+                    rs = null;
+                }
+            } catch(Exception e) {
+                LOG.info("Error in MSSQL Data Provider: "+e.getMessage());
```
Comment (belugabehr) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/MSSQLConnectorProvider.java:69]: `LOG.error("Error in... ", e);`

**Review by belugabehr — CHANGES_REQUESTED** [review]

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java [hunk 8]
```diff
@@ -245,7 +248,12 @@ protected Connection getConnection() {
   private ResultSet fetchTablesViaDBMetaData(String regex) throws SQLException {
     ResultSet rs = null;
     try {
-      rs = getConnection().getMetaData().getTables(scoped_db, null, regex, new String[] { "TABLE" });
+        if (this.type.equalsIgnoreCase(IDataConnectorProvider.ORACLE_TYPE) ||
```
Comment (nrg4878) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java:251]: Sorry if I missed this earlier or if this was recently added, but would it make sense to move this logic into the provider ? so the abstract provider will call something like getCatalog() and getSchema() which will return whatever is appropriate. Maybe the abstract provider can provide the default and providers can override the method. or maybe just add a util method in this class (not abstract) that returns the values for all the DBs.

**Pull request closed without merging**